### ISSUE TITLE
How to Prevent Angular Issues with API Contract Changes Using Zod

### ISSUE THREAD

**Issue by Pasquale-Favella** [Pasquale-Favella/pasquale-favella.github.io#21]
As a front-end developer, one common problem you may face is ensuring your application doesn't break when the API contract changes. Whether you're working with API responses, external data, or handling backend integration, it’s critical to have a solution that validates the data your front-end consumes. In this article, we’ll explore how using `Zod`, a popular data validation library, can help prevent front-end issues due to `API contract changes`.

## Common Problem: Front-End Breaks Due to API Contract Changes

Imagine you’re developing a feature for your project, and you agree on an `API contract` with the back-end team. You write your code, implement the feature, and even cover multiple test scenarios. Everything works perfectly. However, some time later, the feature stops working.

Upon investigation, you find that the `front-end code` is fine, and all your tests are passing. The issue? The `API response` has changed, and the `API contract` you initially agreed upon is no longer valid. As a result, the feature breaks, and these changes propagate through your entire application.

This is a common issue that many developers face when working with `API integrations`. The key question is: how can you prevent your application from breaking due to unexpected changes in external data?

## Solution: Using Zod for Data Validation

One of the best ways to prevent front-end issues caused by API changes is by validating the external data before using it in your application. This is where `Zod`, a lightweight and flexible data validation library, comes into play. [Zod](https://zod.dev/) allows you to define `schemas` for your data and validate it at runtime, ensuring that any incoming data adheres to the expected structure.

Let’s dive into how you can use Zod to solve this issue.

## Step-by-Step Guide: How to Use Zod for Data Validation

Zod supports various types of data validation. Let’s start with something simple: validating string data.

```javascript
import { z } from 'zod';

const StringSchema = z.string();
const EmailSchema = z.string().email();
const LengthSchema = z.string().min(3).max(15);
```
In the above example, we are defining schemas that validate the data type. Zod doesn’t just check if the value is a string but also ensures that the string matches certain conditions—such as validating an email format or checking if the string has a specific length.

Here’s how you can validate data with Zod:

```javascript
StringSchema.parse(456); // Throws an error because 456 is not a string
StringSchema.parse('hello'); // Passes validation because 'hello' is a string

const result = EmailSchema.safeParse('not-an-email'); 
// Returns { success: false, error: {...} }
```
You can also customize the error messages to be more informative:

```javascript
const CustomStringSchema = z.string().refine(val => val === 'expected', {
  message: 'This is not the expected value!',
});
```
## Using Zod to Validate API Responses

Let’s apply Zod to a more complex scenario: `validating API responses`. Imagine you’re working with an API that returns the following structure:

```typescript
interface ApiResponse {
  id: number;
  name: string;
  items: Item[];
}

interface Item {
  productId: string;
  quantity: number;
}
```
In a typical `Angular` application, you might fetch this data using the `HttpClient` service like this:

```typescript
@Injectable({ providedIn: 'root' })
export class ApiService {
  private readonly http = inject(HttpClient);

  fetchData(): Observable<ApiResponse> {
    return this.http.get<ApiResponse>('/api/data');
  }
}
```
Using `Zod`, you can define a schema that mirrors this structure and validates the API response:

```typescript
import { z } from 'zod';

const ApiResponseSchema = z.object({
  id: z.number(),
  name: z.string(),
  items: z.array(
    z.object({
      productId: z.string(),
      quantity: z.number(),
    })
  ),
});

type ApiResponse = z.infer<typeof ApiResponseSchema>;
```
Here’s how you can integrate this `schema` into your service to ensure the data is valid:

```typescript
@Injectable({ providedIn: 'root' })
export class ZodApiService {
  private readonly http = inject(HttpClient);

  fetchData(): Observable<ApiResponse> {
    return this.http.get('/api/data').pipe(
      map((response) => ApiResponseSchema.parse(response)) // Validates the response
    );
  }
}
```
Now, if the API changes unexpectedly—say the backend sends `itemId` instead of `productId`—Zod will throw an error, allowing you to catch the issue immediately. This proactive validation helps prevent bugs caused by inconsistent data structures.

## Advanced Usage: Creating Custom RxJs Operators for Validation

You can simplify the process even further by creating a custom `RxJs operator` to validate data throughout your application:

```typescript
export function validateWithZod<T extends z.ZodTypeAny>(schema: T) {
  return map((response: unknown) => schema.parse(response));
}

@Injectable({ providedIn: 'root' })
export class ZodApiService {
  private readonly http = inject(HttpClient);

  fetchData(): Observable<ApiResponse> {
    return this.http.get('/api/data').pipe(validateWithZod(ApiResponseSchema));
  }
}
```
Now, whenever the backend returns an invalid response, `Zod` will catch it and display a clear error message in your browser’s console.

For instance, if the backend sends the following response:
```json
{
  "id": 101,
  "name": "Example",
  "items": [
    { "productId": "abc", "quantity": 3 },
    { "productId": "def", "quantity": 2 }
  ]
}
```
Zod will validate the response. If the API structure is incorrect, Zod will notify you with an error message, making it clear where things went wrong.

## Why Data Validation is Crucial for Front-End Developers

In modern `front-end development`, applications often rely on data fetched from external APIs. Since this data is external, it falls outside the scope of TypeScript’s type-checking. This means that when the API response structure changes, your front-end application is at risk of breaking—unless you validate the data.

Using a data validation library like Zod allows you to:
- `Ensure data consistency` across your application
- `Catch errors early`, before they propagate through the system
- `Handle unexpected API changes` effectively
- Provide `better error messages` for debugging

## Conclusion

To sum up, `validating API responses` and other external data is a crucial step in ensuring your front-end application remains robust, even when external systems change. While you can generate types from the API, using a runtime validation tool like Zod provides an extra layer of protection. Zod not only helps you validate data but also allows you to catch and handle errors proactively, saving you time and preventing bugs in the long run.

By integrating Zod into your front-end development workflow, you can effectively mitigate the risks associated with changing API contracts and maintain a stable application.

Happy coding!
